### PR TITLE
BUG: Fix merge of a cropped page

### PR DESCRIPTION
**Pull request by pubpub-zz**
tracked in #636

thee smaller box between cropBox is and  trimBox(== mediaBox by default) is used

### CODE REVIEW

**Comment by pubpub-zz** [comment]
@MartinThoma ,
any issue with this PR ?

**Comment by MartinThoma** [comment]
@pubpub-zz I simply didn't have the time to review it so far :sweat_smile:

**Comment by pubpub-zz** [comment]
😳

File: PyPDF2/_page.py [hunk 1]
```diff
@@ -539,16 +539,21 @@ def _merge_page(
         page2content = page2.getContents()
         if page2content is not None:
             page2content = ContentStream(page2content, self.pdf)
+            if(page2.cropBox.getWidth()<page2.trimBox.getWidth() or
+               page2.cropBox.getHeight()<page2.trimBox.getHeight()):
+                rect = page2.cropBox
+            else:
+                rect = page2.trimBox
```
Comment (MartinThoma) [inline PyPDF2/_page.py:546]:
```suggestion
            if (
                page2.cropbox.width < page2.trimbox.width
                or page2.cropbox.height < page2.trimbox.height
            ):
                rect = page2.cropbox
            else:
                rect = page2.trimbox
```

File: tests/test_page.py [hunk 2]
```diff
@@ -185,3 +201,4 @@ def test_page_scale():
 
     assert op.scale(sx=2).ctm == (2, 0, 0, 2, 0, 0)
     assert op.scale(sy=3).ctm == (3, 0, 0, 3, 0, 0)
+
```
Comment (MartinThoma) [inline tests/test_page.py:204]:
```suggestion
```

File: tests/test_page.py [hunk 3]
```diff
@@ -176,6 +177,21 @@ def test_page_rotation_non90():
         page.rotate_clockwise(91)
     assert exc.value.args[0] == "Rotation angle must be a multiple of 90"
 
+
```
Comment (MartinThoma) [inline tests/test_page.py:180]:
```suggestion

```

File: tests/test_page.py [hunk 4]
```diff
@@ -176,6 +177,21 @@ def test_page_rotation_non90():
...
+def test_page_merge_cropped():
+    p = PdfReader(os.path.join(RESOURCE_ROOT, "issue-604.pdf"))
+    a = deepcopy(p.page[1])  # crossed to ease test reading
+    b = deepcopy(p.page[2])
+    a.cropbox = RectangleObject([100, 100, 300, 200])
+    w = PdfWriter()
+    w.add_page(a)
+    w.add_page(b)
+    c = deepcopy(b)
+    c.mergePage(a)
```
Comment (MartinThoma) [inline tests/test_page.py:190]:
```suggestion
    c.merge_page(a)
```

File: tests/test_page.py [hunk 5]
```diff
@@ -176,6 +177,20 @@ def test_page_rotation_non90():
         page.rotate_clockwise(91)
     assert exc.value.args[0] == "Rotation angle must be a multiple of 90"
 
+def test_page_merge_cropped():
+    p = PdfReader(os.path.join(RESOURCE_ROOT, "issue-604.pdf"))
+    a = deepcopy(p.page[1])  # crossed to ease test reading
+    b = deepcopy(p.page[2])
```
Comment (MartinThoma) [inline tests/test_page.py:183]:
```suggestion
    a = deepcopy(p.pages[1])  # crossed to ease test reading
    b = deepcopy(p.pages[2])
```

**Comment by MartinThoma** [comment]
I've just fixed the merge conflicts I've caused. For the review, I still need to do those things:

1. Read #636 and understand the issue
2. Inspect a couple of examples of merges where this change makes a difference

This takes some time. Getting the 2.0.0-dev branch back into main has priority to me as this unblocks other changes. Still, I'm confident that this PR will be handled in the next 4 weeks.

**Comment by MartinThoma** [comment]
I was reading "10.10.1 Page Boundaries":

* The **media box** defines the boundaries of the physical medium on which the page is to be printed. It may include any extended area surrounding the finished page for bleed, printing marks, or other such purposes. It may also include areas close to the edges of the medium that cannot be marked because of physical limitations of the output device. Content falling outside this boundary can safely be discarded without affecting the meaning of the PDF file.
* The **crop box** defines the region to which the contents of the page are to be clipped (cropped) when displayed or printed. Unlike the other boxes, the crop box has no defined meaning in terms of physical page geometry or intended use; it merely imposes clipping on the page contents. However, in the absence of additional information (such as imposition instructions specified in a JDF or PJTF job ticket), the crop box determines how the page’s contents are to be positioned on the output medium. The default value is the page’s media box.
* The **bleed box** (PDF 1.3) defines the region to which the contents of the page should be clipped when output in a production environment. [...] The default value is the page’s crop box.
* The **trim box** (PDF 1.3) defines the intended dimensions of the finished page after trimming. It may be smaller than the media box to allow for production-related content, such as printing instructions, cut marks, or color bars. The default value is the page’s crop box.
* The **art box** (PDF 1.3) defines the extent of the page’s meaningful content (including potential white space) as intended by the page’s creator. The default value is the page’s crop box.

There is also a pretty helpful image.

**Comment by MartinThoma** [comment]
This PR changes the behavior of `_merge_page`. 

It changes which rectangle is appened to the path (`re` operation; TABLE 4.9 Path construction operators).

If the page that is about to be merged has a crop box that is smaller in any dimension than its trim box, the crop box will be added. Otherwise the trim box will be added.

**Comment by MartinThoma** [comment]
@pubpub-zz Why don't we simply always take the cropbox?

**Comment by MartinThoma** [comment]
Pdfjam seems to also use the cropbox, but the positioning is different:

> pdfjam crazyones.pdf git-cropped.pdf --outfile merged-pdfjam.pdf --nup "1x2"

Pdftk also seems to use the cropbox, but the positioning is different from PyPDF2 and pdfjam:

> pdftk git-cropped.pdf background crazyones.pdf output merged-pdftk.pdf

**Comment by pubpub-zz** [comment]
> @pubpub-zz Why don't we simply always take the cropbox?

from PDFRef 1.7, page 145 , CropBox is declared optional. But I've just seen that trimbox is also optional. I've based my proposal on the proposal from the issue.
I agree, we may have also to take into account mediaBox (mandatory field)

**Comment by MartinThoma** [comment]
The part that I'm really confused about is what if the crop box and the trim box are completely disjunct? If they don't overlap at all? What result would you expect?

It seems weird to me to switch from trim box to crop box just because one of the dimensions of the crop box is smaller. What is the logic behind that?

**Comment by MartinThoma** [comment]
By the way, this is how I checked the results:

```python
from PyPDF2 import PdfReader, PdfWriter
from PyPDF2.generic import RectangleObject

reader = PdfReader("box.pdf")
crop_page = reader.pages[0]
print(crop_page.mediabox)
print(crop_page.cropbox)
print(crop_page.trimbox)

writer = PdfWriter()
crop_page.cropbox = RectangleObject((0, 0, 400, 400))
writer.add_page(crop_page)
with open("git-cropped.pdf", "wb") as fp:
    writer.write(fp)

reader = PdfReader("crazyones.pdf")
crazy_page = reader.pages[0]

crazy_page.merge_page(crop_page)

writer = PdfWriter()
writer.add_page(crazy_page)

with open("merged-pypdf2-crop.pdf", "wb") as fp:
    writer.write(fp)
```

File: tests/test_page.py [hunk 6]
```diff
@@ -194,6 +195,20 @@ def test_page_rotation_non90():
         page.rotate(91)
     assert exc.value.args[0] == "Rotation angle must be a multiple of 90"
 
+def test_page_merge_cropped():
```
Comment (MartinThoma) [inline tests/test_page.py:198]:
```suggestion

def test_page_merge_cropped():
```

**Comment by MartinThoma** [comment]
@pubpub-zz I'm sorry that I didn't merge this PR so far. I'm simply really uncertain if this is doing the right thing. The test also currently succeeds (without your change). This makes it hard for me to judge.

Let's make an example:

## Scenario 1

page

Assume the gray part is where we actually have content.

The green rectangle is the cropbox and the red one is the trimbox. What should we use for `rect` and why?

## Scenario 2

grow-2

The green rectangle is the cropbox and the red one is the trimbox. What should we use for `rect` and why?

## Scenario 3

grow-2

The green rectangle is the trimbox and the red one is the cropbox. What should we use for `rect` and why?

**Pull request closed without merging**